### DISCUSSION TITLE
UI-Editing

### DISCUSSION THREAD

**Discussion by bigoulours** [nicotine-plus/nicotine-plus#2507] [Q&A] [answered]
Hi guys,

I'm not familiar with gtk and wanted to learn from your project.
It seems that [Glade](https://gitlab.gnome.org/GNOME/glade) is the currently preferred solution to edit GTK ui-files graphically.

However some files in your repo trigger some warnings/errors when I open (or try to save) them with Glade. Like "Property 'X align' of object class 'Label' was introduced in gtk+ 3.16" in `pynicotine/gtkgui/ui/mainwindow.ui`.

Fair enough, I edited the `required` tag from `<requires lib="gtk+" version="3.0"/>` to `<requires lib="gtk+" version="3.16"/>`. Now I can save without warning but new lines are added by Glade.

So my question: do you edit the xml ui-files manually? If no, what kind of editor do you use?

**Comment by slook** [comment, marked as answer]
Glade is unmaintained and this project's UI files are not optimized for it, so as such many of it's warnings can be ignored. The important Gtk warnings are displayed in the console at runtime (while developing Nicotine+ needs to be runtime tested on gtk3 >= 3.22.30 and gtk4 >= 4.6.8).

> do you edit the xml ui-files manually? 

Glade is sometimes just about usable for initially drafting layouts, but in the end the UI files always need to be edited manually in a text editor or IDE of your choosing that has XML syntax highlighting.

> what kind of editor do you use?

[Geany](https://github.com/geany/geany) is adequate for that purpose.

**Reply by bigoulours** [reply]
Thanks for the clarification!

**Reply by slook** [reply]
It might be interesting to know that the Glade maintainer is now working on "Cambalache is a new RAD tool for Gtk 4 and 3 with a clear MVC design and data model first philosophy"...

- https://gitlab.gnome.org/jpu/cambalache
- https://blogs.gnome.org/xjuan/
- https://www.linuxadictos.com/en/swap-a-tool-for-creating-gtk-user-interfaces.html

It does look more promising than Glade ever was, because the focus seems to be on creating a clean XML data structure. I've not tried it yet.

**Reply by slook** [reply]
If you would like to contribute something to learn on see Issue https://github.com/nicotine-plus/nicotine-plus/issues/1659

**Reply by mathiascode** [reply]
The next version of Cambalache will support our UI files. I'll make changes to our files to ensure they look the same as Cambalache's output (it sorts properties alphabetically).
